### ISSUE TITLE
Improve performance by serving __NEXT_DATA__ in a seperat request

### ISSUE THREAD

**Issue by 0xsven** [zeit/next.js#2795]
Currently I am working on an i18n solution with my nextjs app. In general I don't want to include all the text strings with the initial page load. Especially with huge pages (think of wikipedia or a terms and condition page). All those text strings are already part of the server side rendered document. Why include it again in `__NEXT_DATA__` and add it into the initial page load, when it is not needed for the initial display of the page?

I am not very familiar with all the internals but this goes for a lot of data. Lots of us are using redux. The initial redux store doesn't have to be part of the first load. 

This might not be a huge performance improvement for many websites but for some of them it is. And I don't see any benefit in serving this data with the first load.

What do you think?

**Comment by nikvm** [comment]
I just stumbled upon the same potential issue. I load a post's HTML string in `getInitialProps`. The string is part of the initial render and duplicated in `__NEXT_DATA__`, for re-hydration purposes I guess. I imagine a workflow were said data is loaded in a separate request. And only once this request resolves the app starts hydrating.

**Comment by clintonhalpin** [comment]
Hey Ya'll, I saw this issue so rather than creating a new one I thought I'd tack on my notes on this. Also using Redux.

We are currently rebuilding our website https://quotecatalog.com in Next.js. The issue I'm having is that the document payload is much larger than I'd like.

After some exploration it seems like the root cause is that the app's props are included in `__NEXT_DATA__.props`.

Having the smallest possible document payload is critical for getting a good Lighthouse Score and for users on not great mobile Networks.

### Benchmarks
- Homepage ( https://quotecatalog.com )
  - Props = `~1MB`
  - w/o Props = `204kb`
- Permalink ( https://quotecatalog.com/quote/brianna-wiest-she-wants-you-t-lawdZL1/ )
  - Props = `183kb`
  - w/o Props = `32kb`

(Note these are from our staging server but I included the live URL's for reference, that app is currently in Angular)  

### My Workaround
I got the benchmarks above by creating a custom `NextScript` Component the only difference is that it assigns `__NEXT_DATA__.props` to an empty object.

With this approach the app will render, but once the client side javascript kicks in the pages break because the props aren't there to check. I think this happens in `client/index.js` but I haven't been able to totally figure it out.

You can see the CustomNextScript component [here](https://gist.github.com/clintonhalpin/269f95e0897e91776a51faf03f114725#file-customnextscript-js-L70) but the most important part is that in the NextScript render method I do the following:
```
__NEXT_DATA__.props = {};
```

### Possibly Related issues

1. zeit/next.js#2795
2. zeit/next.js#2389

### Questions

1. Why are props included in the `__NEXT_DATA__` payload?
2. How difficult would this be to fix?
3. Would you accept a PR if I could fix this?

**Comment by krasevych** [comment]
Some one find solution?

**Comment by rashidul0405** [comment]
It’s required for may be re-hydration purpose? Then if we get data via a separate request, performance will degrade? React will have to wait until that request resolves and then hydrate - I doubt whether it’ll gonna work

**Comment by krasevych** [comment]
I have the problem with next export.
I was generated static landing and load translation inside getInitialProps, and then I got the all translation inside __NEXT_DATA__.props.translation. I don't need it, and I wanna remove the translation

**Comment by timneutkens** [comment]
Like @rashidul0405 said, we need the result of `getInitialProps` to re-hydrate React. Without it React won't know what to render on the client side since it doesn't have any data to render with. Feel free to think of another solution for data loading though, I'd be happy to review it.

**Comment by DullReferenceException** [comment]
```
client        | server
--------------|-----------
page request -->  process request
              |   fetch data
              |   render
display page <--  HTML+data
mount w/data  |          
```

```
client           | server
-----------------|-----------
page request    --> process request
display spinner <-- HTML with spinner
mount w/o data   |          
data request    --> process request
                 |  fetch data
ingest data     <-- data
re-render        |
```

**Comment by DOMApe** [comment]
Has anyone figured this out?

**Comment by DullReferenceException** [comment]
@DOMApe: if you want to fetch the data in a separate request, don't use `getInitialProps`, but fetch the data after the component mounts instead. `getInitialProps` is for when you _do_ want the data to be served up with the page payload. It's not a bug. There are legitimate use cases where you want this behavior.

Fetching data takes time. Usually it'll be faster fetching it from the server because the chances are that the data is co-located in the same data center as your app server. If you wait until component mount, you're waiting for a first render (of a spinner/skeleton?), client JS download, CSS parsing, then an XHR request from the client's browser which will be much slower than the server-side request. So although the first option increases _page load time_, it also gives a meaningful render sooner for the customer.

If the data isn't essential for a first meaningful render, you are better off fetching via XHR (not using `getInitialProps`) because in that case the increased page load time isn't as worth it.

**Comment by timneutkens** [comment]
For context this is how instagram.com works: it sets `window._sharedData =`.

On facebook.com you can see `requireLazy`. 

Amazon.com does sort of the same but it's slightly different, it interleaves script tags, I assume because their layout has realtime changes.

Target.com has `window.__PRELOADED_STATE__`.

Youtube.com has `window["ytInitialData"] =`

Yahoo.com has `YUI_config`

Twitter.com has `window.__INITIAL_STATE__`

Next.js, since version 8 uses `<script type="application/json">DATA_HERE</script>` and then parses the script tag before React hydration. I'm not planning on changing this in the near future. 

We might revisit when React Suspense for SSR lands, however even then you need a way to emit the data from server to client.

In case of talking about an initial redux state, in theory it's possible to create a redux store on the server and only pass bits of it for hydration to the client in getInitialProps. 

@DullReferenceException gave a great reply around this so I don't have much to add there 👌

**Comment by lencioni** [comment]
@addyosmani recently [published some advice on the v8 blog about parsing JSON, and recommends using `JSON.parse` with a string literal for optimal speed](https://v8.dev/blog/cost-of-javascript-2019#json).

In response, @HenrikJoreteg [did some profiling of a few different approaches for reading Redux state from the server into the client](https://joreteg.com/blog/improving-redux-state-transfer-performance), and found good performance improvements for using `JSON.parse` over even `<script type="application/json">`.

@timneutkens do you think this approach would have value for next.js?

**Comment by abhijeetdhope** [comment]
Any further development on this issue ?

**Comment by revskill10** [comment]
It's possible to do that. The way is you need to serve your SSR as json as well.
On first request, you just returned normal html, but without the `window.__YOUR_SSR_DATA__ = {}`, then on client side, before hydrating, you make a request to the same url, but with `json`, and then you can use it to hydrate.

**Comment by DullReferenceException** [comment]
> It's possible to do that. The way is you need to serve your SSR as json as well.
> On first request, you just returned normal html, but without the `window.__YOUR_SSR_DATA__ = {}`, then on client side, before hydrating, you make a request to the same url, but with `json`, and then you can use it to hydrate.

Let's say we've got the following definitions:

* R = the overhead of a browser issuing a request
* S = the overhead of the server generating the SSR data
* H = the time it takes to download and process the HTML
* J = the time it takes to download and process the SSR JSON
* TTI = time-to-interactive (the amount of time before the user can interact with the page)

With the current solution of SSR data being included in the HTML, you have an overall TTI of `R+S+H+J`. If you download the JSON in a separate request, the TTI is now `(R+S+H) + (R+S+J)`. The best case scenario (where we emit partial HTML and start to preload the JSON in parallel) would still be `R+S+H+J`, but in reality there will still be a delay before that second request is issued, so it's unlikely this will ever perform better.

**Comment by revskill10** [comment]
@DullReferenceException This works best if you can cache those json data on a separate server. It's clear that the performance is worse for first request only. But you get smaller html size. And you will benefit in case your html has really large json data. I'm still figuring out a better way, but it's currently the way i built my website.

**Comment by teclone** [comment]
In gatsby, gatsby loads initial page props as JSON file. it makes a request for it. To maintain performance, the JSON file is preloaded. just the same way Next JS preloads main script files, it should also preload the associated JSON props file.

This is the best way to achieve a good first meaningful paint. Imagine someone downloading 1MB HTML file vs downloading 200kb HTML. and then download the remaining 800kb JSON file just immediately after receiving the 200kb HTML file.

That is the performance for me. With the current approach, it is hard to achieve good first meaning paint and LCP with NextJS.

If this is not going to be supported, can anyone show how to override the hydration in NEXT.JS? I would be happy to implement a workaround for my project.

**Comment by teclone** [comment]
@DullReferenceException,

But you know that this JSON file can be preloaded, <link rel="preload" as="json" />, so the overhead from the browser perspective can be taken out. then again,
The server does not generate JSON data anymore, since it is already on the server? and even if it was to generate the JSON file it did not exist, it will generate it once, per 1m requests for instance.

This approach is working very well for Gatsby. It does not sound great sending enormous data that is not needed for initial page render.

**Comment by DullReferenceException** [comment]
> It does not sound great sending enormous data that is not needed for initial page render.

If you're generating enormous amounts of data not needed for initial page render, then you're doing it wrong. If you want to defer loading data, then load it after mounting instead of in `getInitialProps` or its ilk.

**Comment by felixmeziere** [comment]
I agree with all of this, for SEO purposes we need to serve an SSGed page with quite a lot of data but currently the architecture of next.js prevents us from doing so with good performance (LCP is pretty bad).

**Comment by felixmeziere** [comment]
> @addyosmani recently [published some advice on the v8 blog about parsing JSON, and recommends using `JSON.parse` with a string literal for optimal speed](https://v8.dev/blog/cost-of-javascript-2019#json).
> 
> In response, @HenrikJoreteg [did some profiling of a few different approaches for reading Redux state from the server into the client](https://joreteg.com/blog/improving-redux-state-transfer-performance), and found good performance improvements for using `JSON.parse` over even `<script type="application/json">`.
> 
> @timneutkens do you think this approach would have value for next.js?

@timneutkens the approach coming from https://joreteg.com/blog/improving-redux-state-transfer-performance suggested by @lencioni is super interesting: there is an 18% TTI improvement out of the box (for a well optimized app, probably more for apps with a lot of data being sent from server to client) obtained by changing 

```html
<script id="__NEXT_DATA__" type="application/json">
  {"props": { ... }}
</script>
```
to
```html
<script>
  window.__NEXT_DATA__" = JSON.parse("{\"props\": { ... }");
</script>
```

it is quite shocking! I would be surprised if the team behind Next.js left this super-low hanging fruit unexploited, have you looked into it?

At the moment we cannot do this in Next even with a "home-made" solution, given how getStaticProps works.

Thanks if you can have a look! :-)

**Comment by ikevinsolomon** [comment]
@felixmeziere holla at me if you're looking for LCP, FCP hack ;) We have been able to lock it down to 1.2 to 1.4s.

**Comment by felixmeziere** [comment]
@ikevinsolomon thanks! Could you explain what the hack consists in?

**Comment by balazsorban44** [comment]
This issue has been automatically locked due to no recent activity. If you are running into a similar issue, please create a new issue with the steps to reproduce. Thank you.
